### PR TITLE
ENG-15117: subselect gives differing results when rerun multiple times

### PR DESCRIPTION
**Pull request by Chaozz**

### CODE REVIEW

File: src/frontend/org/voltdb/planner/StatementPartitioning.java [hunk 1]
```diff
@@ -499,6 +499,15 @@ public void analyzeForMultiPartitionAccess(Collection<StmtTableScan> scans,
                     break;
                 }
             }
+        } else if (scans.size() == 1) {
+            // For query like "select * from (select DL_REGISTER_IP_CITY  from SJYH_DENGLU2 where DL_USER_ID = '113001' ) as result;"
```
Comment (kaikaizi) [inline src/frontend/org/voltdb/planner/StatementPartitioning.java]: Also put JIRA number in here.

File: src/frontend/org/voltdb/planner/StatementPartitioning.java [hunk 2]
```diff
@@ -499,6 +499,15 @@ public void analyzeForMultiPartitionAccess(Collection<StmtTableScan> scans,
...
+            // There is no filter on the outer query,
+            // the inferredParameterIndex should be inferred from the subquery instead of uninitialized (-1).
+            StmtTableScan tableScan = scans.iterator().next();
+            if (tableScan instanceof StmtSubqueryScan) {
```
Comment (kaikaizi) [inline src/frontend/org/voltdb/planner/StatementPartitioning.java]: Should it be an assert? If it fails no other tests, let's change it to assert.

File: src/frontend/org/voltdb/planner/StatementPartitioning.java (hunk 1 above)
Comment (Chaozz) [reply src/frontend/org/voltdb/planner/StatementPartitioning.java]: done.

File: src/frontend/org/voltdb/planner/StatementPartitioning.java (hunk 2 above)
Comment (Chaozz) [reply src/frontend/org/voltdb/planner/StatementPartitioning.java]: No.
`Select * from T` will also hit this branch. In this case the tableScan is a instance of `StmtTargetTableScan`. Only subselect query have the type `StmtSubqueryScan`.

Comment (kaikaizi) [reply src/frontend/org/voltdb/planner/StatementPartitioning.java]: Right. Thanks for the clarification.

**Comment by Chaozz** [comment]
The SQL Coverage tests look good, merge it to master.

**Pull request merged**